hashing/src/lib.rs

```diff
@@ -1,174 +1,165 @@
-    fmt::{self, Debug, Display, Formatter, LowerHex, UpperHex},
-};
+#[cfg(all(feature = "no-std", no_std))]
+use alloc::vec::Vec;
+#[cfg(all(features = "std", not(no_std)))]
+use std::vec::Vec;
 
 use blake2::{
     digest::{Update, VariableOutput},
     VarBlake2b,
 };
-use datasize::DataSize;
-use hex_buffer_serde::{Hex, HexForm};
-use hex_fmt::HexFmt;
-#[cfg(test)]
-use rand::Rng;
-use schemars::JsonSchema;
-use serde::{Deserialize, Serialize};
-
-use casper_execution_engine::shared::newtypes::Blake2bHash;
-use casper_types::bytesrepr::{self, FromBytes, ToBytes};
 
-use super::Error;
-#[cfg(test)]
-use crate::testing::TestRng;
+#[cfg(feature = "std")]
+use hex_buffer_serde::{Hex, HexForm};
 
 /// The hash digest; a wrapped `u8` array.
-#[derive(
-    Copy,
-    Clone,
-    DataSize,
-    Ord,
-    PartialOrd,
-    Eq,
-    PartialEq,
-    Hash,
-    Serialize,
-    Deserialize,
-    Default,
-    JsonSchema,
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Default)]
+#[cfg_attr(
+    feature = "std",
+    derive(
+        schemars::JsonSchema,
+        serde::Serialize,
+        serde::Deserialize,
+        datasize::DataSize
+    ),
+    schemars(with = "String", description = "Hex-encoded hash digest."),
+    serde(deny_unknown_fields)
 )]
-#[serde(deny_unknown_fields)]
-#[schemars(with = "String", description = "Hex-encoded hash digest.")]
 pub struct Digest(
-    #[serde(with = "HexForm::<[u8; Digest::LENGTH]>")]
-    #[schemars(skip, with = "String")]
+    #[cfg_attr(
+        feature = "std",
+        schemars(skip, with = "String"),
+        serde(with = "HexForm::<[u8; Digest::LENGTH]>")
+    )]
     [u8; Digest::LENGTH],
 );
 
 impl Digest {
-    /// Length of `Digest` in bytes.
     pub const LENGTH: usize = 32;
 
-    /// Returns a copy of the wrapped `u8` array.
-    pub fn to_array(self) -> [u8; Digest::LENGTH] {
+    /// Creates a 32-byte hash digest from a given a piece of data
+    pub fn hash<T: AsRef<[u8]>>(data: T) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).expect("should create hasher");
+        hasher.update(data);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
+    }
+
+    /// Hashes a pair of byte slices into a single [`Digest`]
+    pub fn hash_pair<T: AsRef<[u8]>, U: AsRef<[u8]>>(data1: T, data2: U) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).unwrap();
+        hasher.update(data1);
+        hasher.update(data2);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
+    }
+
+    /// Returns a reference to the underlying value.
+    pub fn inner(&self) -> &[u8; Digest::LENGTH] {
+        &self.0
+    }
+
+    /// Returns the underlying value.
+    pub fn value(&self) -> [u8; Digest::LENGTH] {
```

Comment:
```suggestion
    pub fn value(self) -> [u8; Digest::LENGTH] {
```

Reply:
@dwerner when we were investigating this together this appeared to not break anything...